### PR TITLE
Fix Darwin version indentification

### LINKED ISSUE #179: Incorrect(?) darwin version in 1.8 with `Metal.versioninfo()`
I just noticed that when calling `Metal.versioninfo()` with Julia 1.8.5, the Darwin version is reported as `21.5.0`, while in 1.9 it's reported as `22.4.0`. It seems like this behaviour is also found when calling `versioninfo()`. Is this intended behaviour? If not, would a pull request that makes `darwin_version()` use `uname -r` be welcome or do we want to keep the behaviour the same as that of the Julia language?

### PR DESCRIPTION
**Pull request by christiangnrd**
Closes #179 

I went with `uname` instead of the suggestion in issue #179 for simplicity and the fact that either method would only ever get called once per load so speed isn't very important here. 

I'm not opposed to using the other function (and giving credit) if anyone feels strongly about it.

@tgymnich What do you think?

### CODE REVIEW

**Comment by tgymnich** [comment]
You are going for the slower approach but on the other hand you cache the value. Thats a contradiction in my opinion.

**Comment by christiangnrd** [comment]
I'd say this is ready to be merged

File: lib/mtl/MTL.jl [hunk 1]
```diff
@@ -8,10 +8,20 @@ using ObjectiveC, .Foundation, .Dispatch
...
-    machine = Sys.MACHINE
-    VersionNumber(machine[findfirst("darwin", machine)[end]+1:end])
+    if !isassigned(_darwin_version)
+        size = Ref{Csize_t}()
+        err = @ccall sysctlbyname("kern.osrelease"::Cstring, C_NULL::Ptr{Cvoid}, size::Ptr{Csize_t}, C_NULL::Ptr{Cvoid}, 0::Csize_t)::Cint
+        Base.systemerror("sysctlbyname", err != 0)
+        osrelease = Vector{Cchar}(undef, size[])
+        err = @ccall sysctlbyname("kern.osrelease"::Cstring, osrelease::Ptr{Cvoid}, size::Ptr{Csize_t}, C_NULL::Ptr{Cvoid}, 0::Csize_t)::Cint
+        Base.systemerror("sysctlbyname", err != 0)
+        osrelease[end] = 0
```
Comment (maleadt) [inline lib/mtl/MTL.jl:20]: The string already looks null-terminated.

File: lib/mtl/MTL.jl [hunk 2]
```diff
@@ -11,7 +11,14 @@ export darwin_version, macos_version
...
     if !isassigned(_darwin_version)
-        verstr = read(`uname -r`, String)
+        size = Ref{Csize_t}()
+        err = @ccall sysctlbyname("kern.osrelease"::Cstring, C_NULL::Ptr{Cvoid}, size::Ptr{Csize_t}, C_NULL::Ptr{Cvoid}, 0::Csize_t)::Cint
+        Base.systemerror("sysctlbyname", err != 0)
+        osrelease = Vector{Cchar}(undef, size[])
+        err = @ccall sysctlbyname("kern.osrelease"::Cstring, osrelease::Ptr{Cvoid}, size::Ptr{Csize_t}, C_NULL::Ptr{Cvoid}, 0::Csize_t)::Cint
+        Base.systemerror("sysctlbyname", err != 0)
+        osrelease[end] = 0
+        verstr = GC.@preserve osrelease unsafe_string(pointer(osrelease))
```
Comment (maleadt) [inline lib/mtl/MTL.jl]: Note that this is bad use of `GC.@preserve`, as you should preserve across the code that may access `osrelease` unsafely.

Comment (maleadt) [reply lib/mtl/MTL.jl]: Ah oops, no, this is creating an owned string. My mistake.

**Pull request merged**